### PR TITLE
changefeedccl: Add CHANGEFEED privilege

### LINKED ISSUE #85082: changefeedccl: implement granular permissioning model for changefeds
At this point, if an app developer wants to use changefeeds to stream events to an external sink, they need admin access to their CRDB cluster or to request that an admin creates a changefeed for them (bad UX). Granting every application developer admin access increases the overall risk level and adds a data exfiltration threat vector. Instead, data streaming permissions should be more granularly granted to the correct group of users who need them. 

For details, see doc: https://docs.google.com/document/d/108pdrl2ODiLenWYuwBQAkkD0t8lxH9AwyJT54JDVk8w/edit

Jira issue: CRDB-18031

Epic CRDB-14014

### PR DESCRIPTION
**Pull request by HonoreDB**
Adds the CHANGEFEED kind of privilege, which can be used
instead of the CONTROLCHANGEFEED global role option.
    
Release note (sql change): Users may now be granted CHANGEFEED on a table. GRANT CHANGEFEED ON TABLE foo TO max allows max to create changefeeds for foo. Users with the CONTROLCHANGEFEED role option still require the SELECT privilege instead.
    
Informs https://github.com/cockroachdb/cockroach/issues/83843
Fixes https://github.com/cockroachdb/cockroach/issues/85082

Release justification: Additive authorization change needed for compliance.

### CODE REVIEW

**Comment by HonoreDB** [comment]
Pushing for beginning of review but I still need to add logic tests and clean up the privilege check code a bit.

**Comment by HonoreDB** [comment]
Informs https://github.com/cockroachdb/cockroach/issues/83843, @amruss might have a more specific one?

**Comment by HonoreDB** [comment]
Added a more specific issue (thanks Abbey!) and lots more tests. Please take a look!

File: pkg/ccl/changefeedccl/cdctest/testfeed.go [hunk 1]
```diff
@@ -20,6 +20,12 @@ import (
 type TestFeedFactory interface {
 	// Feed creates a new TestFeed.
 	Feed(create string, args ...interface{}) (TestFeed, error)
+
+	// AsUser creates a copy of this TestFeedFactory where all
+	// calls to feed() will be attempted as the given user, and
+	// passes that factory in to fn. An error will only be returned
+	// if the initial connection as the user fails.
+	AsUser(user string, fn func()) error
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/cdctest/testfeed.go:28]: "passes that factory to fn".. but fn is `func()`... This is a bit confusing.

File: pkg/ccl/changefeedccl/alter_changefeed_test.go [hunk 2]
```diff
@@ -1325,3 +1325,45 @@ func TestAlterChangefeedInitialScan(t *testing.T) {
...
+
+		for _, tableName := range []string{`normal`, `sandbox`, `secure`} {
+			rootDB.Exec(t, fmt.Sprintf(`CREATE TABLE %s (id int primary key)`, tableName))
+			rootDB.Exec(t, fmt.Sprintf(`INSERT INTO %s VALUES (0)`, tableName))
+			rootDB.Exec(t, fmt.Sprintf(`GRANT SELECT ON TABLE %s TO guest`, tableName))
+		}
+
+		rootDB.Exec(t, `GRANT CHANGEFEED ON TABLE sandbox TO guest`)
+
+		// TODO: Currently necessary to have CHANGEFEED on preexisting tables.
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/alter_changefeed_test.go:1345]: I think so? Why wouldn't we want to?

File: pkg/ccl/changefeedccl/changefeed_stmt.go [hunk 3]
```diff
@@ -1223,3 +1213,99 @@ func failureTypeForStartupError(err error) changefeedbase.FailureType {
...
 }
+
+// checkPrivileges iterates over all the descriptors returned by ResolveTargetsToDescriptors,
+// which includes descriptors for each targeted table and each database containing one or more
+// targeted tables, mixed together in an arbitrary order. We want to honor privileges at either
+// the database level or the table level, so when we hit a missing privilege at the table level
+// we don't know whether that's fatal until we process its parent descriptor.
+func checkPrivileges(
+	ctx context.Context, p singlePrivilegeChecker, descriptors []catalog.Descriptor,
+) error {
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/changefeed_stmt.go:1224]: Should we just use resolver.ResolveExisting and call checkPrivilege for users?
Alternatively, the same backup package exposts NewResolver method -- that one seems pretty easy to use

**Review by HonoreDB — COMMENTED** [review]
Updated this a bit after meeting on requirements.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85005#-:-N8_BHxU6e3sdSzrOV66:bc9oi8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @gh-casper, @miretskiy, and @samiskin)

___
*[`pkg/ccl/changefeedccl/changefeed_stmt.go` line 1224 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85005#-N7pyeHfBF7LCFWS0LxR-r1-1224:-N8AHbau5fm_LdDfgdR6:bl8pixi) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc06acb7d4780fdf6c31bc816bb5b941dff78817/pkg/ccl/changefeedccl/changefeed_stmt.go#L1224)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Should we just use resolver.ResolveExisting and call checkPrivilege for users?
Alternatively, the same backup package exposts NewResolver method -- that one seems pretty easy to use 
</blockquote></details>

There's probably a way to simplify this using ResolveExisting but I'm not seeing it at the moment--I don't want to just do `checkPrivilege(table) || checkPrivilege( ResolveExisting(table.GetParentID()))` because I want to cache the privilege check on the db level in case we're creating a changefeed on 1000 tables in the same database as admin, and if I'm caching anyway then I might as well skip the "hitting cache" part and use the unique database descriptors I have from the resolver already.
___
*[`pkg/ccl/changefeedccl/cdctest/testfeed.go` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85005#-N7pyeHfBF7LCFWS0LxQ-r1-28:-N8AJJJ84C8XCIF3C0xg:bsxndhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc06acb7d4780fdf6c31bc816bb5b941dff78817/pkg/ccl/changefeedccl/cdctest/testfeed.go#L28)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

"passes that factory to fn".. but fn is `func()`... This is a bit confusing.
</blockquote></details>

Oops, forgot to update the comment after I changed the flow. Fixed.
___
*[`pkg/ccl/changefeedccl/alter_changefeed_test.go` line 1345 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85005#-N7r3txcJ9foPQf-1RH9-r1-1345:-N8AKZZz89FGWDWyxdfz:b-7ngd1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc06acb7d4780fdf6c31bc816bb5b941dff78817/pkg/ccl/changefeedccl/alter_changefeed_test.go#L1345)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think so? Why wouldn't we want to?
</blockquote></details>

(We ended up discussing this in a meeting and deciding this was correct, so I'm removing the TODO.)

**Comment by miretskiy** [comment]
> There's probably a way to simplify this using ResolveExisting but I'm not seeing it at the moment--I don't want to just do `checkPrivilege(table) || checkPrivilege( ResolveExisting(table.GetParentID()))` because I want to cache the privilege check on the db level in case we're creating a changefeed on 1000 tables in the same database as admin, and if I'm caching anyway then I might as well skip the "hitting cache" part and use the unique database descriptors I have from the resolver already.

I don't know... I'd err on the side of readability; the use case of 1k table doesn't come up often; even if it does, perhaps
the priv check system should cache things.. Also, you don't need to do "check(t) || check(parent)"... All you need to do
is to record (unique) parent IDs for tables where check(t) is false, and then just check parents.

**Comment by HonoreDB** [comment]
>All you need to do
is to record (unique) parent IDs for tables where check(t) is false, and then just check parents.

That's what I'm already doing...I guess that's evidence that my code indeed has a readability problem.

**Comment by miretskiy** [comment]
> > All you need to do
> > is to record (unique) parent IDs for tables where check(t) is false, and then just check parents.
> 
> That's what I'm already doing...I guess that's evidence that my code indeed has a readability problem.

It's just from last time I looked at it, it seemed that there was fair amount of code for this.
Few things: perhaps move this priv check function into separate file...
If you're happy w/ current impl, let me know -- and I'll re-review.

**Comment by HonoreDB** [comment]
bors try

**Comment by adityamaru** [comment]
@HonoreDB apologies for all the conflicts, and we're probably going to confict again because of https://github.com/cockroachdb/cockroach/pull/86918. Mine should be ready to bors on green but I can hold off if you want to get yours in first and have me rebase on top.

**Comment by HonoreDB** [comment]
No worries, you go first, looks like I still have some logic tests to update here anyway.

**Comment by HonoreDB** [comment]
Closing in favor of https://github.com/cockroachdb/cockroach/pull/87490

**Pull request closed without merging**